data/transactions/logic/eval.go

```diff
@@ -0,0 +1,1193 @@
+}
+func opByteConstLoad(cx *evalContext) {
+	n := uint(cx.program[cx.pc+1])
+	opByteConstN(cx, n)
+	cx.nextpc = cx.pc + 2
+}
+func opByteConst0(cx *evalContext) {
+	opByteConstN(cx, 0)
+}
+func opByteConst1(cx *evalContext) {
+	opByteConstN(cx, 1)
+}
+func opByteConst2(cx *evalContext) {
+	opByteConstN(cx, 2)
+}
+func opByteConst3(cx *evalContext) {
+	opByteConstN(cx, 3)
+}
+
+func opArgN(cx *evalContext, n uint64) {
+	if n >= uint64(len(cx.Txn.Lsig.Args)) {
+		cx.err = fmt.Errorf("cannot load arg[%d] of %d", n, len(cx.Txn.Lsig.Args))
+		return
+	}
+	cx.stack = append(cx.stack, stackValue{Bytes: cx.Txn.Lsig.Args[n]})
+}
+func opArg(cx *evalContext) {
+	n := uint64(cx.program[cx.pc+1])
+	opArgN(cx, n)
+	cx.nextpc = cx.pc + 2
+}
+func opArg0(cx *evalContext) {
+	opArgN(cx, 0)
+}
+func opArg1(cx *evalContext) {
+	opArgN(cx, 1)
+}
+func opArg2(cx *evalContext) {
+	opArgN(cx, 2)
+}
+func opArg3(cx *evalContext) {
+	opArgN(cx, 3)
+}
+
+func checkBnz(cx *evalContext) int {
+	offset := (uint(cx.program[cx.pc+1]) << 8) | uint(cx.program[cx.pc+2])
+	if offset > 0x7fff {
+		cx.err = fmt.Errorf("bnz offset %x too large", offset)
+		return 1
+	}
+	cx.nextpc = cx.pc + 3
+	target := cx.nextpc + int(offset)
+	if target >= len(cx.program) {
+		cx.err = errors.New("bnz target beyond end of program")
+		return 1
+	}
+	cx.branchTargets = append(cx.branchTargets, target)
+	sort.Ints(cx.branchTargets)
+	return 1
+}
+func opBnz(cx *evalContext) {
+	last := len(cx.stack) - 1
+	cx.nextpc = cx.pc + 3
+	isNonZero := cx.stack[last].Uint != 0
+	cx.stack = cx.stack[:last] // pop
+	if isNonZero {
+		offset := (uint(cx.program[cx.pc+1]) << 8) | uint(cx.program[cx.pc+2])
+		if offset > 0x7fff {
+			cx.err = fmt.Errorf("bnz offset %x too large", offset)
+			return
+		}
+		cx.nextpc += int(offset)
+	}
+}
+
+func opPop(cx *evalContext) {
+	last := len(cx.stack) - 1
+	cx.stack = cx.stack[:last]
+}
+
+func opDup(cx *evalContext) {
+	last := len(cx.stack) - 1
+	sv := cx.stack[last]
+	cx.stack = append(cx.stack, sv)
+}
+
+func (cx *evalContext) txnFieldToStack(txn *transactions.Transaction, field uint64) (sv stackValue, err error) {
+	err = nil
+	switch field {
+	case 0:
+		sv.Bytes = txn.Sender[:]
+	case 1:
+		sv.Uint = txn.Fee.Raw
+	case 2:
+		sv.Uint = uint64(txn.FirstValid)
+	case 3:
+		sv.Uint = uint64(txn.LastValid)
+	case 4:
+		sv.Bytes = txn.Note
+	case 5:
+		sv.Bytes = txn.Receiver[:]
+	case 6:
+		sv.Uint = txn.Amount.Raw
+	case 7:
+		sv.Bytes = txn.CloseRemainderTo[:]
+	case 8:
+		sv.Bytes = txn.VotePK[:]
+	case 9:
+		sv.Bytes = txn.SelectionPK[:]
+	case 10:
+		sv.Uint = uint64(txn.VoteFirst)
+	case 11:
+		sv.Uint = uint64(txn.VoteLast)
+	case 12:
+		sv.Uint = txn.VoteKeyDilution
+	case 13:
+		sv.Bytes = []byte(txn.Type)
+	case 14:
+		sv.Uint = uint64(txnTypeIndexes[string(txn.Type)])
+	case 15:
+		sv.Bytes = make([]byte, 40)
+		copy(sv.Bytes, txn.XferAsset.Creator[:])
+		binary.BigEndian.PutUint64(sv.Bytes[32:], txn.XferAsset.Index)
```

Comment:
Instead of having a custom byte-encoding here, we should probably define one inside of `userBalance.go`.

Reply:
Actually, it may be possible to just refer to `txn.XferAsset.Index` and have only 8 bytes on the stack.